### PR TITLE
feat(security): functional audit pipeline + standalone query

### PR DESCRIPTION
**Pull request by 2witstudios**
## Summary
- Add `audit()` and `auditRequest()` — dual-write functions that log to both the structured logger and tamper-evident audit DB, with automatic IP/user-agent extraction from request headers
- Add standalone `queryAuditEvents()` function and deduplicate `SecurityAuditService.queryEvents()` to delegate to it
- Update security-audit-coverage gate regex to recognize the new `audit()` / `auditRequest()` API
- Add 28 tests covering all paths (risk-level branching, header fallback chains, override-wins, DB failure resilience, query filters, empty header normalization, limit validation)

## Review feedback addressed
- **Empty header normalization** (Codex + CodeRabbit): Header values are now pre-extracted, trimmed, and normalized via `(value || undefined)` so the `??` fallback chain correctly falls through for empty/whitespace strings
- **Limit validation** (CodeRabbit): `limit: 0` is now correctly applied; negative/non-integer limits throw a clear error

## Test plan
- [x] 107/107 audit module tests pass
- [x] Full monorepo build succeeds (13/13 tasks)
- [x] `auditRequest()` correctly extracts IP from `x-forwarded-for` (first entry), falls back to `x-real-ip`, then `"unknown"`
- [x] Empty/whitespace headers correctly fall through to next fallback
- [x] Pre-set `ipAddress`/`userAgent` on the event are not overridden by headers
- [x] `audit()` logs at warn level when `riskScore >= 0.5`, info otherwise
- [x] DB write failures are caught and logged without throwing
- [x] Coverage gate recognizes `audit(` and `auditRequest(` patterns
- [x] `queryAuditEvents({ limit: 0 })` correctly applies limit
- [x] Negative/non-integer limits throw clear error

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added audit logging and request-aware audit functions with automatic persistence.
  * Added audit-event query API to retrieve and filter historical security audit records.
* **Tests**
  * New test suites covering audit logging, request enrichment, query behavior, and resilience on write failures.
  * Updated security-audit coverage check to recognize additional audit call patterns.

### CODE REVIEW

**Review by 2witstudios — COMMENTED** [review]
# 🔬 Code Review: feat(security): functional audit pipeline + standalone query

## Epic Scope Analysis

**Branch**: `pu/audit-foundation`  
**Commit**: `179e0ca4` — feat(security): add functional audit pipeline and standalone query  
**Changed**: 8 files, +520 / -48 lines

This PR introduces two new primitives (`audit()`, `auditRequest()`) and extracts `queryAuditEvents()` as a standalone function. These form the foundation for the 248-route audit coverage epic described in `tasks/security-audit-coverage.md`.

---

## 1. Code Structure & Organization ✅

**Good decisions:**
- New files are cleanly scoped: `audit-log.ts` (write path), `audit-query.ts` (read path) — clear separation of concerns
- Tests colocated in `__tests__/` per project convention
- `index.ts` barrel updated; `server.ts` re-exports added for downstream consumers
- `SecurityAuditService.queryEvents()` now delegates to the standalone function, eliminating duplication without breaking existing callers

**File layout:**
```
packages/lib/src/audit/
├── audit-log.ts           (new — audit() + auditRequest())
├── audit-query.ts         (new — queryAuditEvents())
├── __tests__/
│   ├── audit-log.test.ts  (new — 15 tests)
│   └── audit-query.test.ts (new — 8 tests)
├── security-audit.ts      (modified — query logic extracted, imports updated)
└── index.ts               (modified — new exports)
```

---

## 2. Standards Compliance ✅

**TypeScript:**
- No `any` types — all parameters and return types properly typed ✅
- Clean use of `AuditEvent` and `QueryEventsOptions` interfaces from `security-audit.ts` ✅
- Proper nullish coalescing (`??`) for fallback chains ✅

**Naming:**
- `audit()` / `auditRequest()` — concise verbs, clear intent ✅
- `queryAuditEvents()` — descriptive, distinguishes from the class method ✅

**Comments:**
- Module-level docblocks explain *purpose* and *relationship* to existing patterns ✅
- Function-level JSDoc is minimal and useful (describes fire-and-forget behavior, extraction semantics) ✅
- No redundant comments restating what the code does ✅

---

## 3. Test Coverage & Quality ✅

**101/101 tests pass** (8 test files across the audit module).

The 23 new tests (15 for `audit-log`, 8 for `audit-query`) cover:
- Dual-write to logger + DB ✅
- Risk-level branching (info vs warn at 0.5 threshold) ✅
- DB failure resilience (fire-and-forget catch) ✅
- IP extraction chain: `x-forwarded-for` → `x-real-ip` → `"unknown"` ✅
- User-agent extraction with fallback ✅
- Override-wins semantics (pre-set event fields not clobbered by headers) ✅
- All query filter combinations (userId, eventType, resourceType, resourceId, ipAddress, timestamps, limit) ✅

**TDD compliance:**
- Tests use `describe()` for unit-under-test, `it()` with `given..should` in description ✅
- Test isolation: `vi.clearAllMocks()` in `beforeEach`, no shared mutable state ✅
- Hoisted mocks pattern used correctly for `vi.mock` factories ✅

---

## 4. Security — OWASP Top 10 Scan

| # | OWASP Category | Status | Notes |
|---|----------------|--------|-------|
| A01 | Broken Access Control | ✅ N/A | No access control changes; query function inherits caller's authorization context |
| A02 | Cryptographic Failures | ✅ Safe | Hash chain in `logEvent()` unchanged; new code doesn't handle crypto |
| A03 | Injection | ⚠️ Note | See finding below |
| A04 | Insecure Design | ✅ Good | Dual-write pattern ensures audit trail even if one sink fails |
| A05 | Security Misconfiguration | ✅ N/A | No config changes |
| A06 | Vulnerable Components | ✅ N/A | No new dependencies |
| A07 | Auth Failures | ✅ N/A | No auth logic modified |
| A08 | Data Integrity Failures | ✅ Good | Fire-and-forget still writes to tamper-evident hash chain |
| A09 | Logging & Monitoring | ✅ Good | This PR *improves* audit logging |
| A10 | SSRF | ✅ N/A | No outbound requests |

### A03 — Injection (Minor, Informational)

`queryAuditEvents()` uses Drizzle ORM's `eq()`, `gte()`, `lte()` — these are parameterized by the ORM, so **no SQL injection risk** in practice. However, the function accepts arbitrary `options` without any input validation or sanitization at the function boundary. Since this is an internal library function (not directly exposed to user input), the risk is effectively zero. The trust boundary is at the API route layer where callers validate inputs. **No action needed**, but worth noting for completeness.

---

## 5. Architecture & Design Decisions ✅

**Dual-write pattern is sound:**
- Structured logger for real-time observability (ELK/Datadog)
- DB write for tamper-evident forensic trail
- Fire-and-forget: DB failures logged as warnings, never block the caller
- Matches existing patterns (`auditAuthEvent`, `auditSecurityEvent`, `auditSafe`)

**`auditRequest()` request metadata extraction:**
- Standard IP extraction: `x-forwarded-for` (first entry) → `x-real-ip` → `"unknown"` ✅
- Override semantics correct: caller-provided values win over header extraction ✅

**Query extraction:**
- `queryAuditEvents()` standalone → `SecurityAuditService.queryEvents()` delegates to it
- This is the right direction: standalone functions are easier to use from route handlers than instantiating/initializing a class

---

## 6. Findings

### ⚠️ Finding 1: `audit()` silently drops the `catch` promise (Low Risk)

```typescript
// audit-log.ts:20-25
securityAudit.logEvent(event).catch((error: unknown) => {
  loggers.security.warn('[Audit] audit write failed', { ... });
});
```

The returned Promise from `.catch()` is not stored or awaited — this is intentional (fire-and-forget), and the test suite validates the catch path. However, this means:
- In test environments, the unhandled promise resolution happens asynchronously (the test correctly uses `vi.waitFor()` to handle this)
- In production, the catch will fire but `audit()` returns `void` synchronously

**Verdict**: This is *by design* and matches the existing `auditSafe()` and `auditAuthEvent()` patterns. No change needed. ✅

### ⚠️ Finding 2: `x-forwarded-for` trusts the first entry (Informational)

```typescript
// audit-log.ts:36
request.headers.get('x-forwarded-for')?.split(',')[0]?.trim()
```

This takes the leftmost IP, which is standard when behind a trusted proxy (Caddy). If Caddy is misconfigured or absent, a malicious client could spoof this header. This is standard web infrastructure behavior — the fix is at the proxy layer, not here.

**Verdict**: Correct for the deployment architecture (Caddy reverse proxy). No change needed. ✅

### ℹ️ Finding 3: New API not yet consumed by any route handler

`audit()` and `auditRequest()` are exported but `grep` shows zero usage in `apps/web/src/app/api/`. The coverage gate regex was updated to recognize them, but no route actually calls them yet.

**Verdict**: Expected — this is a foundation PR. The task plan (`tasks/security-audit-coverage.md`) shows 9 follow-up PRs that will wire these into routes. ✅

### ℹ️ Finding 4: Overlap with existing audit adapters

The codebase now has four ways to write audit events:
1. `securityAudit.logEvent()` — class method, DB only
2. `auditAuthEvent()` / `auditSecurityEvent()` — adapter, DB + logger
3. `auditSafe()` — fire-and-forget wrapper around a promise
4. `audit()` / `auditRequest()` — **new**, DB + logger (dual-write)

Functions 2 and 4 have overlapping responsibilities. The new `audit()` is cleaner (no event-type mapping indirection), but the old adapters remain in use.

**Recommendation**: Consider documenting a migration path or deprecation plan for the old adapters. Not blocking for this PR.

---

## 7. Quality Metrics

| Metric | Score |
|--------|-------|
| Tests passing | 101/101 ✅ |
| TypeScript | Clean (0 errors) ✅ |
| New `any` types | 0 ✅ |
| Dead code | None ✅ |
| Unused imports | None ✅ |
| Requirements adherence | Fully met ✅ |

---

## Final Assessment

### Score: 95/100 (Excellent)

| Category | Rating |
|----------|--------|
| Requirements adherence | ✅ 100% |
| Code quality | ✅ 95% |
| Test coverage | ✅ 100% |
| Architecture | ✅ 95% |
| Security | ✅ 100% |

### Blocking Issues: None

### Non-blocking Suggestions:
1. **Plan adapter deprecation** — document when `auditAuthEvent`/`auditSecurityEvent`/`auditSafe` will be replaced by `audit()`/`auditRequest()` (can be a separate issue)

### Recommendation: **APPROVE** ✅

Clean, well-tested foundation for the audit coverage epic. The dual-write pattern is sound, tests are thorough, and the extraction of `queryAuditEvents()` properly deduplicates without breaking existing callers. Ready to merge and build on.

File: packages/lib/src/audit/audit-log.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Dual-write an audit event with automatic request metadata extraction.
+ * Extracts ipAddress and userAgent from the Request headers unless
+ * already provided on the event.
+ */
+export function auditRequest(request: Request, event: AuditEvent): void {
+  const ipAddress =
+    event.ipAddress ??
+    request.headers.get('x-forwarded-for')?.split(',')[0]?.trim() ??
+    request.headers.get('x-real-ip') ??
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/audit/audit-log.ts]: Fixed in 837336f3. Header values are now extracted into local variables, trimmed, and converted to `undefined` when empty via `(forwardedFor || undefined)` so the `??` fallback chain correctly falls through to `x-real-ip` and then `"unknown"`. Added 3 tests covering empty and whitespace-only header edge cases. Thanks for catching this!

File: packages/lib/src/audit/audit-log.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    'unknown';
+
+  const userAgent =
+    event.userAgent ??
+    request.headers.get('user-agent') ??
+    'unknown';
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/audit/audit-log.ts:47]: Fixed in 837336f3 using the exact approach from the suggested fix. Header values are now pre-extracted, trimmed, and normalized via `(value || undefined)` so the `??` fallback chain works correctly for empty/whitespace strings. Added 3 tests: empty `x-forwarded-for`, whitespace-only `x-forwarded-for`, and empty `user-agent`. Thanks!

File: packages/lib/src/audit/audit-query.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  const baseQuery = db
+    .select()
+    .from(securityAuditLog)
+    .where(conditions.length > 0 ? and(...conditions) : undefined)
+    .orderBy(desc(securityAuditLog.timestamp));
+
+  if (options.limit) {
+    return baseQuery.limit(options.limit);
+  }
```
[the remark this answers is not in the document]
Comment (2witstudios) [reply packages/lib/src/audit/audit-query.ts]: Fixed in 837336f3. Changed `if (options.limit)` to `if (options.limit !== undefined)` so `limit: 0` is correctly applied. Also added validation that throws `'limit must be a non-negative integer'` for negative or non-integer values. Added 3 tests: `limit: 0`, negative limit, and non-integer limit. Thanks for the catch!
